src/resumable-upload.ts

```diff
@@ -718,13 +718,37 @@ export class Upload extends Writable {
     };
 
     // If using multiple chunk upload, set appropriate header
-    if (multiChunkMode && expectedUploadSize) {
-      // The '-1' is because the ending byte is inclusive in the request.
-      const endingByte = expectedUploadSize + this.numBytesWritten - 1;
-      headers['Content-Length'] = expectedUploadSize;
+    if (multiChunkMode) {
+      // We need to know how much data is available upstream to set the `Content-Range` header.
+      const oneChunkIterator = this.upstreamIterator(expectedUploadSize, true);
+      const {value} = await oneChunkIterator.next();
+
+      const bytesToUpload = value!.chunk.byteLength;
+
+      // Important: we want to know if the upstream has ended and the queue is empty before
+      // unshifting data back into the queue. This way we will know if this is the last request or not.
+      const isLastChunkOfUpload = !(await this.waitForNextChunk());
+
+      // Important: put the data back in the queue for the actual upload iterator
+      this.unshiftChunkBuffer(value!.chunk);
+
+      let totalObjectSize = this.contentLength;
+
+      if (typeof this.contentLength !== 'number' && isLastChunkOfUpload) {
+        // Let's let the server know this is the last chunk since
+        // we didn't know the content-length beforehand.
+        totalObjectSize = bytesToUpload + this.numBytesWritten;
+      }
+
+      // `- 1` as the ending byte is inclusive in the request.
+      const endingByte = bytesToUpload + this.numBytesWritten - 1;
+
+      // `Content-Length` for multiple chunk uploads is the size of the chunk,
+      // not the overall object
+      headers['Content-Length'] = bytesToUpload;
       headers[
         'Content-Range'
-      ] = `bytes ${this.offset}-${endingByte}/${this.contentLength}`;
+      ] = `bytes ${this.offset}-${endingByte}/${totalObjectSize}`;
```

Comment:
this def looks like a bug in that the content-length may not match the totalObjectSize; I'm guessing this is how this issue came up?

Reply:
Alternatively, you don't need to supply `totalObjectSize` and could instead use `*` and only use `totalObjectSize` once it's known to support streams of unknown byte length.

Reply:
Unfortunately it errors when we use `*` on the final chunk in a multiple chunk transfer (see this commit https://github.com/googleapis/nodejs-storage/pull/2074/commits/53aa25dc69dd0304cdd8b8ae0bdd2dd6bf4d0fed). 

Which makes sense for 2 reasons:
- the final chunk could  be smaller than the minimum chunk size threshold ([256 KiB](https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload))
- it's unclear how the server would know this is the last chunk and we're done uploading

Reply:
Why not buffer up to multiple of 256kib?

Reply:
We do, using the user-provided `chunkSize` property. The last chunk in the upload maybe be less than that (e.g. uploading a 257KiB file in 256KiB chunks).

Reply:
interesting, I would expect the following:

Chunk 1: 256KiB
Chunk 2 (Also Finalize) 1KiB

Does this not work as expected or am I missing context?

Reply:
Unfortunately that does not work. The server doesn't know the 2nd chunk is the last chunk with `*` as the end range for multi-chunk uploads (single chunk upload is fine with `*`).

Reply:
See this error:
![Screen Shot 2022-09-21 at 5 02 45 PM](https://user-images.githubusercontent.com/10792787/191865042-cfbc7e19-e368-4c45-8cfe-eb53a9e5b40d.png)

Reply:
ah, I was saying you need to swap the `*` with the complete object size on the final chunk otherwise you'll get that error as expected.